pytest_fixtures/api_fixtures.py

```diff
@@ -47,34 +55,45 @@ def module_smart_proxy(module_location):
     return entities.SmartProxy(id=smart_proxy.id).read()
 
 
+@pytest.fixture(scope='module')
+def default_domain(module_org, module_location, default_smart_proxy):
```

Comment:
What is the intent here with the use of 'default' for some of these module scoped fixtures?

I think it introduces ambiguity - is it something that exists by default on the satellite test instance? Is it something provided by default in the build, or that robottelo adds to the instance that is a default for the session?

Reply:
These are:

- Either, added to the instance by automation as pre-requisites before the pytest session. Like the domain/ subnet etc.
- Or, the defaults in satellite, like `x86_64` and `i386` are default architectures present in satellite. Same for default OS as `RedHat`.

Reply:
Noted - I think if we're going to use 'default' for fixture names in robottelo, it should be clear what a 'default' is, and I think its very ambiguous to use it for both of these situations.

To me, defaults provided by the build itself in satellite/foreman/etc, shouldn't necessarily be fixtures - there is no setup involved.

For things that the test session is going to use as a default, but are expected pre-requisites in the satellite instances under test, I think fixtures are appropriate to do the lookup - as they can skip the test if the expected entity isn't pre-configured, or can take action to configure it for the session.

Reply:
default fixtures are fixtured so that we will minimize the count/time of reading those fixtures (in short caching them).  There are also 

> For things that the test session is going to use as a default, but are expected pre-requisites in the satellite instances under test, I think fixtures are appropriate to do the lookup - as they can skip the test if the expected entity isn't pre-configured, or can take action to configure it for the session.

Do you prefer to have a fixture that will look for default(pre-requisite) first else create new or raise pytest.failure ?

Reply:
I'd prefer having it lookup and create, only failing if its unable to create.

Reply:
@mshriver Actually we are giving a choice here! One can choose existing or create new! In tests, sometimes we need existing entity and sometimes we need to create new. By lookup and create, we will force users to use existing entity which is not the case for every test.

So I will leave it up to the QE:
1. To use existing -OR-
2. To use newly created -OR-
3. To check the existing and and create new

This can be handled from tests itself instead we forcing from fixture.

Reply:
We're talking about two different use cases within the tests then.

The entities which we are calling 'default', and which we expect to exist for the session and not be destroyed by the tests, should have a lookup+create pattern. This would allow robottelo to run on an instance pre-configured for test (what we could configure for our test targets internally) or on a satellite instance that's just been installed with no further configuration.

Tests which require a unique entity that they will be modifying or destroying should not use the default fixture, but a module or function scoped fixture that is separate from the default fixture.  We could control this with indirect parametrization, but I think its more clear to have a separate fixture.  Some of these would be defined local to the test module if they aren't relevant to other test modules, but fixtures which are relevant to multiple components should be defined globally, but still be separate from the 'default' component fixtures.